woocommerce/woocommerce-ios
# Add notification service extension and framework (without any logic)

Part of #5032 
More details in p91TBi-68e-p2

## Why

To support multi-store push notifications, we want to modify notification content from the server payload in client side so that previous app versions are not affected. To achieve this, we are using [iOS notification service extension](https://developer.apple.com/documentation/usernotifications/modifying_content_in_newly_delivered_notifications). This PR added a notification service extension **WooCommerceNotificationServiceExtension**, and an Xcode framework **NotificationServiceExtension** that contains the main logic for unit testing because XCTest is not supported in app extensions. Since diffs from project files are huge already, this PR only includes the framework setup and skeleton files without any logic to modify notification content.

![NotificationServiceExtension](https://user-images.githubusercontent.com/1945542/136488517-ae8335b4-6555-4cc8-97cc-7d2ea07df6fb.jpg)

## Changes

📝 Apology for the big PR! Lots of diffs are on the project files from framework changes. Feel free to review commit by commit to see the important changes in project files and Info.plists. You can also check out the documentation changes first.

- Added notification service extension called `WooCommerceNotificationServiceExtension` in https://github.com/woocommerce/woocommerce-ios/commit/7dcf8fe8b63c32662d01c3cee0b953f167ed5245, it's a target under the WooCommerce project
  - This extension depends on `NotificationServiceExtension` framework, and just has one class `NotificationService: UNNotificationServiceExtension`. Even though the framework below also has the same subclass where our logic will reside, we can't directly use the class from the framework in this app extension. I tried setting `NotificationServiceExtension.NotificationService` in the Info.plist's `NSExtensionPrincipalClass` field but it does not work (the code in the framework won't be called)
- Added an Xcode framework called `NotificationServiceExtension` in https://github.com/woocommerce/woocommerce-ios/commit/b80997c24d3b109e0ee1d7e961d2d5b27af9b37a, which `WooCommerceNotificationServiceExtension` depends on.
  - This framework depends on `Experiments` framework from https://github.com/woocommerce/woocommerce-ios/pull/5106, and it tests target depends on `TestKit`
- Added/updated documentation

## Testing

Prerequisites: the logged-in user has more than one site connected to the app.
If you are testing with your a8c WP.com account, you can resend previous new order/review notifications from the **MC Push Notifications Debug Console** (MC > Push Notifications > Debug Console > WP.com Notification, or with this path `/mobile/push-notifications/debug/?type=wpcom-notification`).

Please test push notifications on a device, since simulators cannot register for remote notifications from WP.com. You might need a developer sandbox for receiving push notifications in a debug build. If you don't have one, please check out the first step in p99K0U-1qz-p2

Let's make sure the push notifications work as before:

- Launch the app from PR branch
- Put the app in the background
- Place an order on any connected site (you can also resend a notification in the MC Debug Console if you're using a8c WP.com account) --> a new order push notification should show up, and the content is the same as before (`You have a new order! 🎉`)
- Leave a review on any connected site --> a new review push notification should show up, and the content is the same as before (`{{customer name}} left a review on {{product name}}`)

Update release notes:
